zebra/rtadv.c

```diff
@@ -493,8 +493,8 @@ static int rtadv_timer(struct thread *thread)
 
 	RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id)
```

Comment:
Why do we go though all the VRFs here in the first place?
If we only want to process interfaces that are in this `zvrf`, then we need to remove this `RB_FOREACH` loop and go only through interfaces in `zvrf->vrf`.

Reply:
No I think this is only applicable to netns-vrfs. RAs under vrf-lite happens in all vrfs because RA's aren't vrf-lite delimited but we use the default vrf's socket. So under vrf-lite, we need to walk the entire list.

Reply:
Got it. But then the change is still not right as it basically does the same thing as working only with one VRF.

Reply:
From my test, the thread is per vrf. In my test, I have two vrfs (default 0 and ppi 1),  I added a print in side the loop, and got the following log:

```
2021/10/15 07:53:06 ZEBRA: [ZQ5HQ-0HZF7] Bali rtadv_timer: ifp->[name ipb0, index: 6, vrf_id: 0], zvrf->[name: default, vrf_id:0] <<<<<<<< zvrf id 0
2021/10/15 07:53:06 ZEBRA: [H1BFG-MHSRG] Fast RA Rexmit on interface ipb0(default:6)
2021/10/15 07:53:06 ZEBRA: [JY2SR-KEYTH] ipb0(default:6): Tx RA, socket 15 <<<<< success

..

2021/10/15 07:53:06 ZEBRA: [ZQ5HQ-0HZF7] Bali rtadv_timer: ifp->[name ipb0, index: 6, vrf_id: 0], zvrf->[name: ppi, vrf_id:1]
2021/10/15 07:53:06 ZEBRA: [H1BFG-MHSRG] Fast RA Rexmit on interface ipb0(default:6)
2021/10/15 07:53:06 ZEBRA: [JY2SR-KEYTH] ipb0(default:6): Tx RA, socket 21 <<<<<
2021/10/15 07:53:06 ZEBRA: [QGWPP-XPTHX][EC 100663299] ipb0(6): Tx RA failed, socket 21 error 19 (No such device) <<<< failed.
```

And you can see the interface ipb0 belongs to the default vrf-0, so the second RA sent failed because it uses the vrf-1's socket.

so I agree we don't need to go through all the VRFs in the RB_FOREACH(), just loop the interfaces in the zvrf's.

Reply:
@leibaogit please do the same test with the VRF-lite backend. As @sworleys says, when using VRF-lite, there's only one thread and one socket in the default VRF, so in this case we actually need to go through all VRFs. It looks like we need to do two different things for netns and VRF-lite here.

Reply:
@idryzhov ok, I can do the test, but can you guiding me how to use the VRF-lite, any documents can be provided? I'm actually new to FRR, thanks.

Reply:
VRF-lite is a linux kernel construct:
https://www.kernel.org/doc/html/latest/networking/vrf.html

After you create the vrfs in the kernel with iproute2 or ifupdown, we can use them in FRR without any special startup arguments. (So remove the `-n` from your daemons file and netns stuff from frr.conf)

Our docs are here:
http://docs.frrouting.org/en/latest/zebra.html?highlight=vrf#virtual-routing-and-forwarding

The cumulus docs here are also the same (using the linux commands):
https://docs.nvidia.com/networking-ethernet-software/cumulus-linux-44/Layer-3/VRFs/Virtual-Routing-and-Forwarding-VRF/

Reply:
frr is supposed to support netns based vrfs, so there should be smth to do.

Reply:
Yea but what this patch does also shouldn't break vrf-lite which as it stands, I believe will since we only use the default vrf in vrf-lite for the RA socket.

To be clear, under vrf-lite, that `zvrf` struct we are matching against with this patch will only ever be the default VRF so all interfaces not in it will be broken.

With netns, we schedule the timers on all the vrfs since they need their own sockets. You can see this code in `ipv6_nd_suppress_ra_set()`, we call `zvrf = rtadv_interface_get_zvrf(ifp);` to get the zvrf we set the thread on. Under vrf-lite this will always return the default.

Reply:
@sworleys I thinking that, if using the vrf-lite, then all interfaces will be in one vrf, right? If so, then all the interfaces'->vrf_id will be the same as the zvrf->vrf_id, then this patch still works, for the vrf-lite, right?

Reply:
@leibaogit No, they aren't in the VRF, the just use the default VRF's socket.

if `eth1` is in `vrf-red` then the `ifp->vrf_id` will point to `vrf-red` not the default vrf BUT it needs to use the default vrf socket for RAs.

this check here: `ifp->vrf_id != zvrf->vrf->vrf_id` will fail for any vrf not in the default which is wrong for vrf-lite RAs.